### PR TITLE
feat(site): Smooth scrolling on nav link

### LINKED ISSUE #237: Smooth Scroll for Navigation
Currently, the navigation jumps abruptly to a new section.  To create a better user experience, implement a smoother scroll to each of the individual sections.

- [x] Research the best method to implement a smooth scroll

- [ ] Implement the smooth scroll

- [ ] Test (include https://gist.github.com/csteinert/6c02cc6f17329c754e15fc4e2e762dc6 in PR)

https://css-tricks.com/smooth-scrolling-accessibility/

**Notes**
Current issues with the implementation (https://github.com/fisshy/react-scroll) (branch: 237-smooth-scroll):
- Scrolling links do not have proper hover cursor styles
- Global Nav uses the same markup on the Blog pages, but the scroll links only work if they're on the same page as the linked content (so you can't navigate back)
- The home page jumps back to the top any time the mobile nav panel is opened on small screens.

### PR DESCRIPTION
**Pull request by zachcd**
### feat(site): Smooth scrolling on nav link use
* Add smooth scrolling package that applies to all anchor links
* Change top link to a #top anchor tag

note: the polyfill for this has a conflict with Gatsby SSR, so I'm importing the non-polyfilled script directly.  Browsers without support will stick to jump navigation instead of scroll.

smoothscrollb2io

closes #237

### CODE REVIEW

**Comment by zachcd** [comment]
looks like there's a deploy issue with the package I used.  Something to work on.

**Comment by FrankyMartz** [comment]
What about the native `<element>.scrollIntoView()`? It looks like it has decent support and a Polyfill.

https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView

**Polyfill:**
http://iamdustan.com/smoothscroll/

**Comment by zachcd** [comment]
@FrankyMartz I'm using smooth-scroll for easing, speed, and now offset purposes.  Just a little more control over the animation.

@tborres looks like the nav menu on mobile locks scrolling while its open.  Shouldn't be too hard to fix.

**Comment by zachcd** [comment]
Bringing in Issues and PRs: #251 #261 #282 #285 

The Mobile Nav was resetting scroll position on open and preventing scrolling while open so as to make sure that touch scrolling wouldn't accidentally happen on the background viewport while in the mobile nav.  This disabling of scrolling prevented the smooth-scroll from happening. 

In order to keep the same functionality but allow programmatic scrolling(and solve the resetting of scroll position): I set touch-action: none, see: https://developer.mozilla.org/en-US/docs/Web/CSS/touch-action

File: src/components/molecules/GlobalNavigation.jsx [hunk 1]
```diff
@@ -3,6 +3,7 @@ import styled from 'styled-components';
 import PropTypes from 'prop-types';
 import { em } from 'polished';
 import windowSize from 'react-window-size';
+import SmoothScroll from 'smooth-scroll/dist/smooth-scroll'
```
Comment (tborres) [inline src/components/molecules/GlobalNavigation.jsx]: ❓ Couldn't you import directly from the npm package name?

```suggestion
import SmoothScroll from 'smooth-scroll';
```

Also, missing a semi-colon. With the right plugins and settings you can setup auto-fix on save so that most of these common misses are auto-fixed.  

**VS Code Extensions**
- ESLint (2.1.1)
- Prettier (3.20.0)

**Settings:**
```
"editor.formatOnSave": true,
"editor.codeActionsOnSave": {
    "source.fixAll.eslint": true
  },
```

File: src/components/molecules/GlobalNavigation.jsx [hunk 2]
```diff
@@ -19,6 +20,16 @@ const mobileBottomMenu = {
   value: `${em('75px')}`,
 };
 
+if (typeof window !== `undefined`) {
+  SmoothScroll('a[href*="#"]',
```
Comment (tborres) [inline src/components/molecules/GlobalNavigation.jsx]: 🎨 Our eslint/prettier rules would typically put the opening `{` on this line.

File: src/components/molecules/GlobalNavigation.jsx [hunk 3]
```diff
@@ -19,6 +20,16 @@ const mobileBottomMenu = {
   value: `${em('75px')}`,
 };
 
+if (typeof window !== `undefined`) {
+  SmoothScroll('a[href*="#"]', 
+  {
+    easing: 'easeOutQuad', 
+    speed: 700,
+    speedAsDuration: true,
+    offset: '50'
```
Comment (tborres) [inline src/components/molecules/GlobalNavigation.jsx]: Prefer alphabetical order of properties to reduce the chance of duplicates and ease future merges (though if you look around you can see that this rule isn't always followed).

File: src/components/molecules/GlobalNavigation.jsx [hunk 4]
```diff
@@ -34,6 +45,7 @@ const NavBar = styled.nav`
   &.menu-open {
     height: 100vh;
     transition: height 300ms ease-in;
+    touch-action: none;
```
Comment (tborres) [inline src/components/molecules/GlobalNavigation.jsx]: Alphabetical order

File: src/pages/index.jsx [hunk 5]
```diff
@@ -32,7 +32,7 @@ injectGlobal`
   }
 
   .noScroll > div {
-    height: 100vh;
+    height: 100%;
```
Comment (tborres) [inline src/pages/index.jsx]: ❓ Why is this change needed?

File: src/components/molecules/GlobalNavigation.jsx (hunk 1 above)
Comment (zachcd) [reply src/components/molecules/GlobalNavigation.jsx]: The import from the dist is because the package references window, so when Gatsby imports it to complete a build, it crashes.  This skips the package's wrapper that references window.  If there's a more "gatsby"-like solution, I'd be very happy to hear about it.

Comment (tborres) [reply src/components/molecules/GlobalNavigation.jsx]: @mewdriller do you know of a more gatsby-like approach to this window issue that Zach ran into?

File: src/pages/index.jsx (hunk 5 above)
Comment (zachcd) [reply src/pages/index.jsx]: 100vh was shortening the height of the div when the menu opened, resetting scrolling position.  This is a bugfix that I'm adding in that fixes the jump to the top on hamburger menu open bug.

Setting the height is still important because of the overflow setting. 
```In order for overflow to have an effect, the block-level container must have either a set height (height or max-height) or white-space set to nowrap.```

Comment (zachcd) [reply src/pages/index.jsx]: It was needed before, because the 100vh was resetting the scroll position and making it jump to the top, 100% was fixing that issue.  I can just delete this whole style now as we don't need it with the use of touch-action.

File: src/components/molecules/GlobalNavigation.jsx (hunk 1 above)
Comment (zachcd) [reply src/components/molecules/GlobalNavigation.jsx]: https://www.gatsbyjs.org/docs/using-client-side-only-packages/

They don't have anything specific, and their first recommendation is to "work around the problem"
